### PR TITLE
Substitute SQL parameters using `unprepare` when converting to native

### LINKED ISSUE #3549: When converting a QB query to SQL, prepared statement params aren't substituted
We don't substitute prepared statements parameters when converting to SQL, e.x.

``` sql
SELECT count(*) AS "count"
FROM "PUBLIC"."PRODUCTS"
WHERE "PUBLIC"."PRODUCTS"."CATEGORY" = ?
```

This results in a somewhat confusing `Parameter "#1" is not set` error.

Ideally we should substitute the actually parameters, but if we can't do that we should provide a more helpful warning of some kind.

### PR DESCRIPTION
**Pull request by jonbender**
Fixes #3549 by calling unprepare in the generic SQL driver when converting queries to native.

I tested locally with H2 and postgres and `lein test`.  Going to see if CI picks up anything in engines I didn't test.

###### TODO 
-  [ x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

**Comment by camsaul** [comment]
@jonbender this looks good I'll run this against CI

**Comment by camsaul** [comment]
@jonbender I've opened PR #5257 to get it to run against CI (we've disabled automatically running external PRs against CI so shady people can't open PRs that would dump out the environment variables containing connection info to our test DBs)

Keep an eye on that. If all the tests pass then this should be mergeable

**Comment by jonbender** [comment]
@camsaul Thanks for taking a look.  Is there any documentation on how to get a full fledged test harness setup locally?  I'd just as soon run the full suite of engines locally before putting up a PR for things like this.

**Comment by camsaul** [comment]
@jonbender you'll need Oracle, Redshift, BigQuery, a Druid cluster, and SQL Server running somewhere, most likely externally. You'll need Postgres, MySQL, and Mongo running somewhere too, but these are fairly easy to run locally. Vertica and Presto you'd need to run locally as well but you can use Docker for those.

So the short answer is it's not easy to get a full test setup locally but if can take a look in `./bin/ci` to see how we get tests running and what sort of env vars to pass in for the various DBs

**Comment by EvaSchreyer** [comment]
Hi, the issue that filters are empty or contain "?" when switching from metabase query builder to SQL still happens in v0.27.

File: src/metabase/driver/generic_sql/util/unprepare.clj [hunk 1]
```diff
@@ -14,11 +14,18 @@
...
+                              (first (hsql/format
+                                      (cond
+                                        (fn? iso-8601-fn) (iso-8601-fn date-exp)
+                                        (keyword? iso-8601-fn) (hsql/call iso-8601-fn (hx/literal date-exp))
+                                        :otherwise (hx/literal date-exp)))))))
 
 (defn unprepare
   "Convert a normal SQL `[statement & prepared-statement-args]` vector into a flat, non-prepared statement."
-  ^String [[sql & args] & {:keys [quote-escape iso-8601-fn], :or {quote-escape "\\\\", iso-8601-fn :timestamp}}]
+  ^String [[sql & args] & {:keys [quote-escape iso-8601-fn], :or {quote-escape "\\\\", iso-8601-fn nil}}]
```
Comment (camsaul) [inline src/metabase/driver/generic_sql/util/unprepare.clj]: @jonbender, can you fill me in on why you made the default implementation for this `nil`? It breaks a few databases so just wondering which ones it fixed

**Review by camsaul — CHANGES_REQUESTED** [review]
I just realized this after reviewing this PR a bit further. 

This code path modified by this PR is actually used for every query that gets ran. The changes in this PR thus eliminate the use of prepared statement arguments (i.e., `?` args) entirely, including for queries that the user never wants to convert to SQL. 

Removing the prepared statement args and passing things in directly by splicing them into the SQL opens up potential SQL injection and type-handling bugs. At any rate, if we did want to take that approach, it makes little sense to put `?` placeholders in the SQL and then splice corresponding SQL snippets in their place afterwards, rather than skipping the `?` step and inserting the snippets directly when we generate the SQL.

I'm no longer sure there's an easy fix for this issue; this will require a bit of thinking on our part. We'll probably want to change how the query processor works so we can generate both a query keeping the prepared statement param placeholders (`?`s) and a second version of the query (no `?`s) suitable for use when a user clicks on the "convert to native" button. Alternatively, we could consider adding an API endpoint to generate the latter when requested. Either way, this will require frontend changes and new tests to make sure this works as intended. 

@jonbender thanks for all your hard work exploring possible solutions to this issue. I'll confer with the Metabase team and when we come up with a solution to this issue I hope we can incorporate some of your contributions from this PR

**Comment by jonbender** [comment]
Thanks for the detailed writeup. It all makes sense, the idea of retaining the original SQL along with the "unprepared" version had occurred to me when I started, but I thought we could skate by just pushing through the unprepared flavor. I agree it might make sense to add a separate call and only unprepare on demand too.

I'll close this one since it won't really cover what you need for #3549

**Pull request closed without merging**